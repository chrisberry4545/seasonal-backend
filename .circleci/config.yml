version: 2
jobs:
  test:
    docker:
      - image: circleci/node:8.10

    steps:
      - checkout

      - setup_remote_docker

      - run:
          name: Output dir
          command: echo $PWD
      - run:
          name: Run Docker
          command: docker-compose run app npm test

      - run:
          name: Run tests
          command: docker exec app npm run test

  deploy:
    docker:
      - image: circleci/node:8.10

    steps:
      - checkout

      - restore_cache:
          keys:
          - v1-dependencies-
          # fallback to using the latest cache if no exact match is found
          - v1-dependencies-

      - run:
          name: Copy env config
          command: |
            cp .env-ci.yml .env.yml

      - run:
          name: Install Serverless CLI and dependencies
          command: |
            sudo npm i -g serverless
            npm install

      - run:
          name: Run tests
          command: npm test

      - run:
          name: Build dist version
          command: |
            npm run build
            npm run move-dist-to-top-level

      - save_cache:
          paths:
            - node_modules
          key: v1-dependencies-

      - run:
          name: Deploy application
          command: sls deploy --stage prod

workflows:
  version: 2
  ci:
    jobs:
      - deploy:
          filters:
            branches:
              only:
                - master
      - test
