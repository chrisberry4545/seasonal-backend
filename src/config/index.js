module.exports = {
  AIRTABLE_API_KEY: process.env.AIRTABLE_API_KEY,
  SKIP_CACHE: process.env.SKIP_CACHE || false
};
