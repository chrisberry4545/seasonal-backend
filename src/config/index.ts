export const SKIP_CACHE = process.env.SKIP_CACHE || false;
export const SEASON_ENDPOINT =
  process.env.SEASON_ENDPOINT || 'season';
export const FOOD_ENDPOINT =
  process.env.FOOD_ENDPOINT || 'food';
export const SEASON_WITH_FOOD_ENDPOINT =
  process.env.SEASON_WITH_FOOD_ENDPOINT || 'season-with-food';
export const SEASON_WITH_RECIPES_ENDPOINT =
  process.env.SEASON_WITH_RECIPES_ENDPOINT || 'season-with-recipes';
export const COUNTRY_ENDPOINT =
  process.env.COUNTRY_ENDPOINT || 'country';
export const V2_ENDPOINT = process.env.V2_ENDPOINT || 'v2';
export const DB_CONNECTION_STRING = process.env.DB_CONNECTION_STRING;

export const DEFAULT_COUNTRY_ID =
  process.env.DEFAULT_COUNTRY_ID || '83533319-9b7a-4c79-9261-4c02bc6b4d0d';
