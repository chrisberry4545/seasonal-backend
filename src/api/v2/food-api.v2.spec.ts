import {
  app
} from '../../app';

import supertest, { Response } from 'supertest';
import { V2_ENDPOINT, FOOD_ENDPOINT } from '../../config';
import {
  RECIPES_ID_PICKLED_BEETROOT,
  SEASON_ID_JANUARY,
  SEASON_ID_FEBRUARY,
  FOOD_ID_BEETROOT,
  FOOD_ID_ONION,
  RECIPES_ID_APPLE_CHEESE_AND_ONION
} from './shared-test-ids';

const v2FoodUrl = `${V2_ENDPOINT}/${FOOD_ENDPOINT}`;

const makeSingleFoodRequest = async (
  id: string = FOOD_ID_BEETROOT,
  isVegetarian?: boolean,
  isVegan?: boolean
) => {
  const query = [
    isVegetarian && 'is-vegetarian=true',
    isVegan && 'is-vegan=true'
  ].filter(Boolean).join('&');
  const queryString = query ? `?${query}` : '';
  return supertest(app).get(`/${v2FoodUrl}/${id}${queryString}`);
};

describe('Get single food item', () => {
  let response: Response;
  beforeAll(async () => {
    response = await makeSingleFoodRequest();
  });

  test('Returns a status of 200', () => {
    expect(response.status).toBe(200);
  });
  test('Retrieves a single food item', () => {
    expect(response.body).toMatchSnapshot();
  });
  test('Populates a foods seasons', () => {
    expect(response.body.seasons.length > 0).toBe(true);
  });
  test('Hydrates the first seasons name', () => {
    expect(response.body.seasons[0].name).toBe('January');
  });
  test('Hydrates the second seasons name', () => {
    expect(response.body.seasons[1].name).toBe('February');
  });
  test('Hydrates the first seasons seasonIndexes', () => {
    expect(response.body.seasons[0].seasonIndex).toBe(0);
  });
  test('Hydrates the second seasons seasonIndexes', () => {
    expect(response.body.seasons[1].seasonIndex).toBe(1);
  });
  test('Hydrates the seasons ids of the first season', () => {
    expect(response.body.seasons[0].id).toBe(SEASON_ID_JANUARY);
  });
  test('Hydrates the seasons ids of the second season', () => {
    expect(response.body.seasons[1].id).toBe(SEASON_ID_FEBRUARY);
  });
  test('Hydrates a foods primaryFoodInRecipe ids if they exist', () => {
    expect(response.body.primaryFoodInRecipe[0].id).toBe(RECIPES_ID_PICKLED_BEETROOT);
  });
  test('Hydrates a foods primaryFoodInRecipe names if they exist', () => {
    expect(response.body.primaryFoodInRecipe[0].name).toBe('Pickled Beetroot');
  });
  test('Returns an empty array for a foods secondaryFoodInRecipe if there is none', () => {
    expect(response.body.secondaryFoodInRecipe).toHaveLength(0);
  });

  describe('When the isVegetarian filter is applied', () => {
    beforeEach(async () => {
      response = await makeSingleFoodRequest(FOOD_ID_BEETROOT, true);
    });

    test('Returns a status of 200', () => {
      expect(response.status).toBe(200);
    });
    test('Filters out non vegetarian/vegan from the primaryFood recipes', () => {
      expect(response.body.primaryFoodInRecipe).toHaveLength(1);
    });
    test('Filters out non vegetarian/vegan from the secondaryFood recipes', () => {
      expect(response.body.primaryFoodInRecipe).toHaveLength(1);
    });
    test('Returns the vegan recipes', () => {
      expect(response.body.primaryFoodInRecipe[0].isVegan).toBe(true);
    });
  });

  describe('When the isVegan filter is applied', () => {
    beforeEach(async () => {
      response = await makeSingleFoodRequest(FOOD_ID_BEETROOT, false, true);
    });

    test('Returns a status of 200', () => {
      expect(response.status).toBe(200);
    });
    test('Filters out non vegan recipes', () => {
      expect(response.body.primaryFoodInRecipe).toHaveLength(1);
    });
    test('Filters out vegetarian recipes', () => {
      expect(response.body.secondaryFoodInRecipe).toHaveLength(0);
    });
    test('Returns only the vegan recipes', () => {
      expect(response.body.primaryFoodInRecipe[0].isVegan).toBe(true);
    });
  });

  describe('when the food is a secondary food item in recipes', () => {
    beforeAll(async () => {
      response = await makeSingleFoodRequest(
        FOOD_ID_ONION
      );
    });

    test('Returns a status of 200', () => {
      expect(response.status).toBe(200);
    });
    test('Retrieves a single food item', () => {
      expect(response.body).toMatchSnapshot();
    });
    test('Retrieves a foods secondaryFoodInRecipe names if they exist', () => {
      expect(response.body.secondaryFoodInRecipe[0].id)
        .toBe(RECIPES_ID_APPLE_CHEESE_AND_ONION);
    });
    test('Retrieves a foods secondaryFoodInRecipe names if they exist', () => {
      expect(response.body.secondaryFoodInRecipe[0].name).toBe('Apple, Cheese & Onion');
    });

    describe('and the isVegan recipe is applied', () => {
      beforeAll(async () => {
        response = await makeSingleFoodRequest(
          FOOD_ID_ONION, false, true
        );
      });

      test('returns the expected result', () => {
        expect(response.body).toMatchSnapshot();
      });
      test('filters out non vegan recipes', () => {
        expect(response.body.secondaryFoodInRecipe).toHaveLength(0);
      });
    });

    describe('and the isVegetarian recipe is applied', () => {
      beforeAll(async () => {
        response = await makeSingleFoodRequest(
          FOOD_ID_ONION, true
        );
      });

      test('returns the expected result', () => {
        expect(response.body).toMatchSnapshot();
      });
      test('does not filter out vegetarian recipes', () => {
        expect(response.body.secondaryFoodInRecipe).toHaveLength(1);
      });
    });
  });
});
