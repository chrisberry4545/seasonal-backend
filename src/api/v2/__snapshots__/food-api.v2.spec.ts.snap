// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get single food item Retrieves a single food item 1`] = `
Object {
  "id": "c6f78568-fe23-47e4-8e65-55934199a39f",
  "imageUrlSmall": "https://beetroot.com",
  "name": "Beetroot",
  "primaryFoodInRecipe": Array [
    Object {
      "id": "0f7fb18e-ac1d-4023-b315-91ca7e29ce4a",
      "imageUrlSmall": "https://pickled-beetroot.com/image",
      "isVegan": true,
      "isVegetarian": false,
      "linkUrl": "https://pickled-beetroot.com",
      "name": "Pickled Beetroot",
    },
    Object {
      "id": "053c46cd-584e-46c5-a70d-b79b9ded3cca",
      "imageUrlSmall": "https://apple-beetroot-and-cheese.com/image",
      "isVegan": false,
      "isVegetarian": false,
      "linkUrl": "https://apple-beetroot-and-cheese.com",
      "name": "Apple, Beetroot & Meat",
    },
  ],
  "seasons": Array [
    Object {
      "id": "41bcbb3a-5599-4f0c-8da9-3144cc5be6de",
      "name": "January",
      "seasonIndex": 0,
    },
    Object {
      "id": "eec0b2bb-4fbd-46df-b905-8d2ee2fb991a",
      "name": "February",
      "seasonIndex": 1,
    },
  ],
  "secondaryFoodInRecipe": Array [],
}
`;

exports[`Get single food item when the food is a secondary food item in recipes Retrieves a single food item 1`] = `
Object {
  "id": "d1df368e-6ed1-4a50-bb7d-6163c15df1d3",
  "imageUrlSmall": "https://onion.com",
  "name": "Onion",
  "primaryFoodInRecipe": Array [],
  "seasons": Array [],
  "secondaryFoodInRecipe": Array [
    Object {
      "id": "fefc209f-ddaf-465d-941f-fe3fefa6b931",
      "imageUrlSmall": "https://apple-and-cheese.com/image",
      "isVegan": false,
      "isVegetarian": true,
      "linkUrl": "https://apple-and-cheese.com",
      "name": "Apple, Cheese & Onion",
    },
  ],
}
`;

exports[`Get single food item when the food is a secondary food item in recipes and the isVegan recipe is applied returns the expected result 1`] = `
Object {
  "id": "d1df368e-6ed1-4a50-bb7d-6163c15df1d3",
  "imageUrlSmall": "https://onion.com",
  "name": "Onion",
  "primaryFoodInRecipe": Array [],
  "seasons": Array [],
  "secondaryFoodInRecipe": Array [],
}
`;

exports[`Get single food item when the food is a secondary food item in recipes and the isVegetarian recipe is applied returns the expected result 1`] = `
Object {
  "id": "d1df368e-6ed1-4a50-bb7d-6163c15df1d3",
  "imageUrlSmall": "https://onion.com",
  "name": "Onion",
  "primaryFoodInRecipe": Array [],
  "seasons": Array [],
  "secondaryFoodInRecipe": Array [
    Object {
      "id": "fefc209f-ddaf-465d-941f-fe3fefa6b931",
      "imageUrlSmall": "https://apple-and-cheese.com/image",
      "isVegan": false,
      "isVegetarian": true,
      "linkUrl": "https://apple-and-cheese.com",
      "name": "Apple, Cheese & Onion",
    },
  ],
}
`;
